src/cli/commands/upgrade.js

```diff
@@ -21,12 +34,95 @@ export function hasWrapper(commander: Object, args: Array<string>): boolean {
 export const requireLockfile = true;
 
 export async function run(config: Config, reporter: Reporter, flags: Object, args: Array<string>): Promise<void> {
-  const useLockfile = args.length || flags.latest;
-  const lockfile = useLockfile ? await Lockfile.fromDirectory(config.lockfileFolder, reporter) : new Lockfile();
-  const {dependencies, devDependencies, optionalDependencies, peerDependencies} =
-    (await config.readRootManifest()) || {};
-  const allDependencies = Object.assign({}, peerDependencies, optionalDependencies, devDependencies, dependencies);
-  let addArgs = [];
+  const lockfile = await Lockfile.fromDirectory(config.lockfileFolder);
+  const deps = await getOutdated(config, reporter, flags, lockfile, args);
+
+  // if specific versions were requested for packages, override what getOutdated reported as the latest to install
+  args.forEach(requestedPattern => {
+    const normalized = PackageRequest.normalizePattern(requestedPattern);
+    const newPattern = `${normalized.name}@${normalized.range}`;
+    let found = false;
+
+    deps.forEach(dep => {
+      if (normalized.hasVersion && dep.name === normalized.name) {
+        found = true;
+        dep.latestPattern = newPattern;
+      }
+    });
+
+    if (normalized.hasVersion && !found) {
+      deps.push({
+        name: normalized.name,
+        wanted: '',
+        latest: '',
+        url: '',
+        hint: '',
+        range: '',
+        current: '',
+        latestPattern: newPattern,
+      });
+    }
```

Comment:
This was the case where a user specifies an exact version for something that isn't outdated. For example if you have foo@2.0.0 and that happens to be the latest, it won't be returned by getOutdatedPackages, but the user can still do `yarn upgrade foo@1.2.4` and we would change to that requested version.
This is preserving functionality that the previous `upgrade` had (because it would just forward the package with version through to `add`)